pkgs/servers/http/nginx/nix-etag-1.15.4.patch

```diff
@@ -0,0 +1,87 @@
+From 3ed81ef9e973247d0c95bf240b3bd2e640f2605a Mon Sep 17 00:00:00 2001
+From: Yegor Timoshenko <yegortimoshenko@riseup.net>
+Date: Fri, 28 Sep 2018 03:27:04 +0000
+Subject: [PATCH] If root is in Nix store, set ETag to its path hash
+
+---
+ src/http/ngx_http_core_module.c | 56 +++++++++++++++++++++++++++------
+ 1 file changed, 47 insertions(+), 9 deletions(-)
+
+diff --git a/src/http/ngx_http_core_module.c b/src/http/ngx_http_core_module.c
+index c57ec00c..d996d7e3 100644
+--- a/src/http/ngx_http_core_module.c
++++ b/src/http/ngx_http_core_module.c
+@@ -1583,6 +1583,7 @@ ngx_http_set_etag(ngx_http_request_t *r)
+ {
+     ngx_table_elt_t           *etag;
+     ngx_http_core_loc_conf_t  *clcf;
++    u_char                    *real, *ptr1, *ptr2;
+ 
+     clcf = ngx_http_get_module_loc_conf(r, ngx_http_core_module);
+ 
+@@ -1598,16 +1599,53 @@ ngx_http_set_etag(ngx_http_request_t *r)
+     etag->hash = 1;
+     ngx_str_set(&etag->key, "ETag");
+ 
+-    etag->value.data = ngx_pnalloc(r->pool, NGX_OFF_T_LEN + NGX_TIME_T_LEN + 3);
+-    if (etag->value.data == NULL) {
+-        etag->hash = 0;
+-        return NGX_ERROR;
+-    }
++    real = ngx_realpath(clcf->root.data, NULL);
```

Comment:
If `clcf->root.data` starts with `/nix/store/`, does this actually mean that the current request is a static file inside that root?  Can you break it using the `alias` directive?  (From my cursory reading of the nginx source code, `proxy_pass` seems to be fine since it bypasses this method entirely.)
```nginx
root /nix/store/...;

location /foo {
  alias /srv/foo;
}
```

Reply:
It works as expected! Here's how to reproduce:

```
$ cat /etc/nixos/configuration.nix
{ pkgs, ... }:

let
  srv = pkgs.runCommand "srv" {} ''
    mkdir $out && echo "<!DOCTYPE html><b>Hello :-)</b>" > $out/index.html
  '';
in

{
  # ...

  services.nginx = {
    enable = true;
    package = pkgs.nginx.overrideAttrs (self: {
      patches = (self.patches or []) ++ [ ./nix-etag-1.15.4.patch ];
    });
    virtualHosts.default = {
      root = srv;
      locations."/foo".alias = "/tmp/srv";
      listen = [({ addr = "127.0.0.1"; port = 8080; })];
    };
  };
}
$ cd /etc/nixos; sudo curl -LO https://raw.githubusercontent.com/NixOS/nixpkgs/971f8e2c779f57a3dfc931161d266b2c32b1d67e/pkgs/servers/http/nginx/nix-etag-1.15.4.patch
$ mkdir -p /tmp/srv
$ echo "<!DOCTYPE html><i>correct</i>" > /tmp/srv/index.html
$ sudo nixos-rebuild switch
$ curl --head http://localhost:8080
HTTP/1.1 200 OK
Server: nginx
Date: Tue, 16 Oct 2018 12:46:29 GMT
Content-Type: text/html
Content-Length: 32
Last-Modified: Thu, 01 Jan 1970 00:00:01 GMT
Connection: keep-alive
ETag: "rxawlg1ivq2zjz9qgfz9anvrwwcj4j9q"
Accept-Ranges: bytes
$ curl --head http://localhost:8080/foo/
HTTP/1.1 200 OK
Server: nginx
Date: Tue, 16 Oct 2018 12:46:41 GMT
Content-Type: text/html
Content-Length: 15
Last-Modified: Tue, 16 Oct 2018 12:43:00 GMT
Connection: keep-alive
ETag: "5bc5dcd4-f"
Accept-Ranges: bytes
```

Meaning, `alias` requests override `root`.